cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,392 @@
+
+
+## Operator changes
+
+Operator will start managing CIDs, and will become the only system pod that issues mutating requests for CIDs to the kube-apiserver. A new controller (CID controller) will run inside Operator that will be responsible for managing CIDs and updating CEPs.
+
+1) Operator requires watching the following resources to allocate CIDs and assign them to CEPs:
+
+* CEPs / Pods (Already watches) - To manage CIDs based on pod labels and update CEPs with correct CIDs.
+* Namespaces - to create and delete CIDs based on namespace labels changes.
+* CIDs - to keep the desired state of CIDs. Required for:
+    * Getting the initial state on start-up.
+    * Being compatible with Agents managing CIDs simultaneously.
+    * Handling unplanned CID changes (create, update, delete) by other sources -- not Operator or Agents.
+
+2) Operator will modify the following resources:
+
+* CIDs - Create and delete. CIDs shouldn’t be modified. If namespace or pod labels are changed, new CIDs will be created for all unique label sets that don’t match any existing identity. Unused CIDs will be cleaned up by the CID GC.
+* CEPs - Update the `Status.Identity` field.
+
+3) CID GC reworked:
+
+* Heartbeat status annotations will no longer be used. It will not be added to CIDs by the Operator, or removed by the Agent. Operator will remain to be compatible with Agents managing CIDs simultaneously. Agents just won’t ever have a chance to remove the annotation.
+* CIDs will still be GC-ed periodically -- every 10 minutes.
+* CIDs will be deleted only when they aren’t used anywhere, in CID, CEP and CES (if enabled) controller’s caches and watcher stores.
+
+4) On startup, the CID controller will cache existing CIDs from the CID watcher store and usage of CIDs in CEPs from the CEP watcher store. The cache will be used for mapping CID name (256 - 65k integer) to security labels and the reverse mapping -- labels to CID name.
+
+5) The CID controller will periodically run CID reconciliation for all CEPs, alongside CID GC, on startup and every 10 minutes. It will make sure that all CEPs have correct CIDs assigned to them.
+
+6) There will be no changes to the CES controller running inside Operator.
+
+1. The CES controller receives CEP updates with assigned CIDs, after the CID controller updates them, and then batch them into CESs.
+2. The CES controller ignores CEPs that have empty (nil) `Status.Identity`. This means that for new CEPs, the CES controller skips CEP creation events that have empty CID, and then batches them on the next immediate update events from the CID controller, where CIDs are assigned.
+
+## Agent changes
+
+Agents will stop managing global security identities (CIDs). Agents will no longer perform following actions:
+
+* Allocate security identities in the global range [256, 65536].
+* Create CIDs. Agents will no longer need the permission to modify CIDs.
+* Remove heartbeat status annotation on CIDs (part of CID GC).
+
+**Temporary ID**
```

Comment:
Thanks for bringing this up. We agreed after the initial review that temporary IDs are not going to be introduced, because, as you said, the complexity outweighs the benefits. We agreed to move forward with the "block ID assignment (including CNI) as long as CID is missing".

I realize that I needed to update this at the top of the document in the "Review discussion" section. We should update it now.

We could remove the "Temporary ID" section or move it to the end of the document to serve as an archived record of considered options during the review.

No part of the implementation that has been merged or in development depends on temp ID, because we resolved it during the design, before the implementation process started.

Reply:
I _think_ it probably makes sense to keep some summary that the idea was considered, what it was, and why we didn't think it was ideal at this time. Broadly speaking it's possible for us to revisit such decisions in future if some of the underlying assumptions shift, so it'd be nice to be able to refer back and explain why we made the decisions we made originally.

One more piece of minor feedback on this: I think it would benefit new readers (and returning readers who have not recently looked) to move the "Review discussion" section later in the document. In the [CFP template](https://github.com/cilium/design-cfps/blob/main/cfps/CFP-003-template.md) that sort of content would typically come under the "Impacts / Key Questions" section in the towards the end. The reasoning just being that it helps the reader to understand if you gradually introduce detail to them. Minor nit and maybe this is a UX issue on GDocs vs. GitHub, but I share it as this did create a little bit of extra friction today for me as a reviewer to look through the proposal, especially since that review section was already discussed and resolved but remains near the top of the document despite no longer being relevant to readers.

Reply:
(I see also that the last point I raised above is moot and you already plan to address that per your comment in the [thread below](https://github.com/cilium/design-cfps/pull/19#discussion_r1711862813))